api/graphql/applications/application_decline_serializers.py

```diff
@@ -0,0 +1,84 @@
+from rest_framework import serializers
+
+from api.graphql.base_serializers import PrimaryKeySerializer
+from applications.models import (
+    Application,
+    ApplicationEvent,
+    ApplicationEventStatus,
+    ApplicationStatus,
+)
+
+
+class ApplicationDeclineSerializer(PrimaryKeySerializer):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.fields["status"].read_only = True
+
+    class Meta:
+        model = Application
+        fields = [
+            "pk",
+            "status",
+        ]
+
+    CAN_DECLINE_STATUSES = (
+        ApplicationStatus.IN_REVIEW,
+        ApplicationStatus.REVIEW_DONE,
+        ApplicationStatus.ALLOCATED,
+    )
+
+    def validate(self, data):
+        if self.instance.status not in self.CAN_DECLINE_STATUSES:
+            raise serializers.ValidationError(
+                f"Only applications with status as {', '.join(self.CAN_DECLINE_STATUSES)} can be approved."
+            )
+        data = super().validate(data)
+
+        return data
+
+    def save(self, **kwargs):
+        for event in self.instance.application_events.all():
+            event.set_status(ApplicationEventStatus.DECLINED)
+        return self.instance
```

Comment:
First I thought that you are missing `self.instance.set_status(ApplicationStatus.DECLINED)`, but then I remembered that this is not how `ApplicationStatus` works 😄